add GLSL_HUAWEI_subpass_shading (#163)
KhronosGroup/GLSL: extensions/huawei/GLSL_HUAWEI_subpass_shading.txt

    from the other shader processors. Compilation units written in the OpenGL Shading Language to run on this
    processor are called subpass shading shaders. When a set of subpass shading shaders are successfully compiled 
    and linked, they result in a subpass shading shader executable that runs on the subpass shading processor.

    A subpass shading shader is almost identical to a compute shader except the following
    differences.
        
        * A subpass shading shader must be dispatched in a dedicated subpass
        * A subpass shading shader is allowed to read fragment froms input attachments
        * The work dimension is defined by the size of render area and the number of layers of the framebuffer
        
Changes to Chapter 4 of The OpenGL Shading Language Specification, Version 4.60

    Modify Section 4.1.7, (Opaque Types)

    Change the "Subpass-input types are only available in fragment shaders." to 

    Subpass-input types are only available in fragment or subpass shading shaders.

Changes to Chapter 8 of The OpenGL Shading Language Specification, Version 4.60 

>     Modify Section 8.18, (Subpass-Input Functions), 

Reviewer: ```suggestion
    Modify Section 8.18, (Subpass-Input Functions)
```

Name

    HUAWEI_subpass_shading

Name Strings

    GL_HUAWEI_subpass_shading

Contact

    Hueilong Wang (wyvernathuawei), HUAWEI

Contributors

    Hueilong Wang, HUAWEI

Status

    Draft

Version

    Last Modified Date: 2021-04-26
    Revision: 1

Dependencies

    This extension can be applied to OpenGL GLSL versions 4.60
    (#version 460) and higher.

    This extension can be applied to OpenGL ES ESSL versions 3.20
    (#version 320) and higher.
    
    This extension is written against revision 7 of the OpenGL Shading Language
    version 4.60, dated July 10, 2019.

>     This extension is written against revision 6 of the OpenGL ES ESSL

Reviewer: I don't think you need to include the actual changes twice, you can just say that the extension can be applied to ES, but that it's written against the desktop version. If there are any significant differences between desktop and ES then you can note them at the end in a section of "Interactions with Open GL ES". Sorry to have misled you on this.

    
Modifications to the OpenGL Shading Language Specification, Version 4.60

    Including the following line in a shader can be used to control the
    language features described in this extension:

      #extension GL_HUAWEI_subpass_shading                          : <behavior>

    where <behavior> is as specified in section 3.3.
    New preprocessor #defines are added:

      #define GL_HUAWEI_subpass_shading  

Additions to Chapter 2 of the OpenGL Shading Language Specification, Version 4.60

    Add Section 2.7. Subpass Shading Processor
    
    The subpass shading processor is a programmable unit that operates the subpass inputs. It runs independently 
    from the other shader processors. Compilation units written in the OpenGL Shading Language to run on this
    processor are called subpass shading shaders. When a set of subpass shading shaders are successfully compiled 
    and linked, they result in a subpass shading shader executable that runs on the subpass shading processor.

    A subpass shading shader is almost identical to a compute shader except the following
    differences.
        

>         * A subpass shading shader must be dispatched in a dedicate subpass

Reviewer: ```suggestion
        * A subpass shading shader must be dispatched in a dedicated subpass
```

>         * A subpass shading shader is allowed to read fragment from input attachment

Reviewer: ```suggestion
        * A subpass shading shader is allowed to read fragments from input attachments
```

        * The work dimension is defined by the size of render area and the number of layers of the framebuffer
        
Changes to Chapter 4 of The OpenGL Shading Language Specification, Version 4.60

    Modify Section 4.1.7, (Opaque Types)

    Change the "Subpass-input types are only available in fragment shaders." to 

    Subpass-input types are only available in fragment or subpass shading shaders.

Changes to Chapter 8 of The OpenGL Shading Language Specification, Version 4.60

    Modify Section 8.18, (Subpass-Input Functions)

    Change the "Subpass-input functions are only available when targeting a Vulkan fragment stage." to 

    Subpass-input functions are only available when targeting a Vulkan fragment or subpass shading stage.

Changes to Chapter 12 of The OpenGL Shading Language Specification, Version 4.60

    Modify Section 12.2.4, (Vulkan Only: Samplers, Images, Textures, and Buffers)

    Change the "Subpass inputs are read through a new set of types, available only to fragment shaders:" to 

    Subpass inputs are read through a new set of types, available only to fragment or subpass shading shaders:

> Additions to Chapter 2 of the OpenGL ES ESSL versions 3.20

Reviewer: As above, I would remove these sections about ES and only note where there are differences from desktop, if any.

Name

    HUAWEI_subpass_shading

Name Strings

    GL_HUAWEI_subpass_shading

Contact

    Hueilong Wang (wyvernathuawei), HUAWEI

Contributors

    Hueilong Wang, HUAWEI

Status

    Draft

Version

    Last Modified Date: 2021-04-26
    Revision: 1

Dependencies

    This extension can be applied to OpenGL GLSL versions 4.60
    (#version 460) and higher.

>     This extension is written against revision 7 of the OpenGL Shading Language

Reviewer: You don't give any interactions with the ES version of the spec. Is it your intention that the ES version of the shading language should be compatible with these shaders?

Author: Okay, I added them in https://github.com/KhronosGroup/GLSL/pull/163/commits/1114f6bed5d92f19481adecdad26a081fa1b4e09

    version 4.60, dated July 10, 2019.

    This extension interacts with revision 43 of the GL_KHR_vulkan_glsl
    extension, dated October 25, 2017.

Overview

>     This extension allows a subpass shading shader to delcare subpass inputs and use subpass input functions.

Reviewer: I find "subpass shading shader" quite awkward. Do you prefer that term to the simpler "subpass shader"?

Author: I've tried but finally I still have to accept it since it pulls the strings to other naming like pipelines, stages.

Reviewer: OK, fair enough.

Modifications to the OpenGL Shading Language Specification, Version 4.60

    Including the following line in a shader can be used to control the
    language features described in this extension:

      #extension GL_HUAWEI_subpass_shading                          : <behavior>

    where <behavior> is as specified in section 3.3.
    New preprocessor #defines are added:

      #define GL_HUAWEI_subpass_shading  

Changes to Chapter 4 of The OpenGL Shading Language Specification, Version 4.60

    Modify Section 4.1.7, (Opaque Types)

    Change the "Subpass-input types are only available in fragment shaders." to 

    Subpass-input types are only available in fragment and subpass shading shaders.

Changes to Chapter 8 of The OpenGL Shading Language Specification, Version 4.60

    Modify Section 8.18, (Subpass-Input Functions)

    Change the "Subpass-input functions are only available when targeting a Vulkan fragment stage." to 

>     Subpass-input functions are only available when targeting a Vulkan fragment and subpass shading stage.

Reviewer: I think this one should be "fragment *or* subpass shading stage"

Author: Thanks, I fixed in https://github.com/KhronosGroup/GLSL/pull/163/commits/1114f6bed5d92f19481adecdad26a081fa1b4e09
